facilitator/src/batch.rs

```diff
@@ -299,84 +300,97 @@ impl<'a, H: Header, P: Packet> BatchWriter<'a, H, P> {
             batch,
             transport,
             trace_id,
+            use_bogus_packet_file_digest: false,
         }
     }
 
     pub fn path(&self) -> String {
         self.transport.path()
     }
 
-    /// Encode the provided header into Avro, sign that representation with the
-    /// provided key and write the header into the batch. Returns the signature
-    /// on success.
-    pub fn put_header(&self, header: &H, key: &EcdsaKeyPair) -> Result<Signature> {
-        let mut transport_writer = self.transport.put(self.batch.header_key(), self.trace_id)?;
-        let mut header_buf = Vec::new();
-        header.write(&mut header_buf)?;
-
-        transport_writer.write_all(&header_buf)?;
-        transport_writer
-            .complete_upload()
-            .context("failed to complete batch header upload")?;
-
-        // TODO(brandon): per the advice of ring's doc comments, create a single SystemRandom &
-        // reuse it throughout the application rather than creating a new SystemRandom each time.
-        key.sign(&SystemRandom::new(), &header_buf)
-            .context("failed to sign header file")
+    /// Sets whether this BatchWriter will use a bogus value for the packet
+    /// file digest when constructing the header of a validation batch. This
+    /// is intended only for testing.
+    pub fn set_use_bogus_packet_file_digest(&mut self, bogus: bool) {
+        self.use_bogus_packet_file_digest = bogus;
     }
 
-    /// Creates an avro_rs::Writer and provides it to the caller-provided
-    /// function, which may then write arbitrarily many packets into it. The
-    /// Avro encoding of the packet will be digested while they are written.
-    /// The operation should return Ok(()) when it has finished successfully or
-    /// some Err() otherwise. packet_file_writer returns the digest of all the
-    /// content written by the operation.
-    pub fn packet_file_writer<F>(&self, operation: F) -> Result<Digest>
-    where
-        F: FnOnce(&mut Writer<&mut DigestWriter<&mut Box<dyn TransportWriter>>>) -> Result<()>,
-    {
+    pub fn write(
+        &self,
+        key: &BatchSigningKey,
+        mut header: H,
+        packets: impl Iterator<Item = Result<P>>,
```

Comment:
I think using [`IntoIterator`](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) would make this a bit more ergonomic. Something like:

```
fn write<I: IntoTerator<Item = Result<P>>(&self, key, header, packets: I) -> Result<()> {
    let packets = packets.into_iter();
    for packet in packets { ... }
}
```

Now the caller can pass in anything that implements `IntoIterator`, so anything from a `Vec[Result<P>]` to `&[Result<P>]` to a hashmap to etc., and we hide the call to `into_iter` here in this method instead of making every caller do it.

[Example of this pattern from `libprio`](https://github.com/abetterinternet/libprio-rs/blob/10e4d168c04546256916d940cde8b4ba9ac432a5/src/vdaf/hits.rs#L142)

I also found some discussion of the tradeoff between making a function generic in `T: Trait` vs. using `impl Trait` [in the Rust book](https://doc.rust-lang.org/reference/types/impl-trait.html#anonymous-type-parameters)

Reply:
Separately, I'm not sure how I feel about `Item = Result<P>`. It seems like this forces most of the callers of `BatchWriter:write` to wrap their packets in `Ok()` to accommodate one callsite where the packet construction can fail. I'm OK with it, but I did want to point you to [this neat feature of `Result`](https://doc.rust-lang.org/rust-by-example/error/iter_result.html). Instead of taking `Iterator<Item = Result<P>>`, you could take `Iterator<Item = P>`, and expect/require callers that have an iterator of `Result<P>` to use a `.collect()` call to map those into `P` while still handling errors. The other advantage is that errors that occur in the calling function no longer get handled here in `write()`.

That being said, I think what you have here works and is good enough for an internal interface, so if you prefer this to putting a call to `collect` with a turbofish in `intake.rs`, that's fine.

Reply:
I like switching to `IntoIter` & switching to `T: Trait` over `impl Trait`. Done.

I slightly preferred using an iterator of `Result<P>` only because it allows us to potentially only parse a single packet at a time, instead of needing to keep all in memory (as is necessary if we collect to a `Vec`). But we don't actually make use of that, and your suggestion is certainly easier to read, so done. (I did a little digging at the realistic packet sizes -- for ingestion packets in `us-ca`, the largest packet file I could find was a few megabytes total -- so I suspect any potential memory savings is unnecessary at the moment.)